File: bridge/whatsapp/bridge.js
/**
 * drbot WhatsApp Bridge
 *
 * WebSocket server that bridges drbot to WhatsApp via Baileys.
 *
 * Protocol:
 * - All messages are JSON with a "type" field
 * - Rust -> Bridge: init, send_message, send_media, get_qr, status, disconnect
 * - Bridge -> Rust: connection, qr, message, sent, error, ready
 */

const { default: makeWASocket, useMultiFileAuthState, DisconnectReason, fetchLatestBaileysVersion } = require('@whiskeysockets/baileys');
const { WebSocketServer } = require('ws');
const pino = require('pino');
const qrcode = require('qrcode-terminal');
const path = require('path');
const fs = require('fs');

const PORT = process.env.PORT || 3001;
const logger = pino({ level: process.env.LOG_LEVEL || 'info' });

let sock = null;
let wsClient = null;
let sessionDir = '.whatsapp';

// Send message to Rust client
function send(msg) {
    if (wsClient && wsClient.readyState === 1) {
        wsClient.send(JSON.stringify(msg));
    }
}

// Initialize WhatsApp connection
async function initWhatsApp() {
    logger.info({ sessionDir }, 'Initializing WhatsApp connection');

    // Ensure session directory exists
    if (!fs.existsSync(sessionDir)) {
        fs.mkdirSync(sessionDir, { recursive: true });
    }

    const { state, saveCreds } = await useMultiFileAuthState(sessionDir);
    const { version } = await fetchLatestBaileysVersion();

    sock = makeWASocket({
        version,
        logger: pino({ level: 'silent' }),
        auth: state,
        printQRInTerminal: false,
        browser: ['drbot', 'Chrome', '120.0.0'],
    });

    // Handle connection updates
    sock.ev.on('connection.update', async (update) => {
        const { connection, lastDisconnect, qr } = update;

        if (qr) {
            logger.info('QR code received');
            // Print to terminal for convenience
            qrcode.generate(qr, { small: true });
            // Send to Rust client
            send({ type: 'qr', qr });
        }

        if (connection === 'close') {
            const statusCode = lastDisconnect?.error?.output?.statusCode;
            const shouldReconnect = statusCode !== DisconnectReason.loggedOut;

            logger.info({ statusCode, shouldReconnect }, 'Connection closed');
            send({ type: 'connection', status: 'close' });

            if (shouldReconnect) {
                logger.info('Reconnecting...');
                setTimeout(initWhatsApp, 3000);
            } else {
                send({ type: 'connection', status: 'loggedOut' });
            }
        } else if (connection === 'open') {
            logger.info('Connected to WhatsApp');
            send({ type: 'connection', status: 'open' });
            send({ type: 'ready' });
        } else if (connection === 'connecting') {
            send({ type: 'connection', status: 'connecting' });
        }
    });

    // Save credentials when they update
    sock.ev.on('creds.update', saveCreds);

    // Handle incoming messages
    sock.ev.on('messages.upsert', async ({ messages, type }) => {
        if (type !== 'notify') return;

        for (const msg of messages) {
            // Skip status updates
            if (msg.key.remoteJid === 'status@broadcast') continue;

            const messageContent = msg.message;
            if (!messageContent) continue;

            // Extract text content
            let text = null;
            if (messageContent.conversation) {
                text = messageContent.conversation;
            } else if (messageContent.extendedTextMessage) {
                text = messageContent.extendedTextMessage.text;
            } else if (messageContent.imageMessage?.caption) {
                text = messageContent.imageMessage.caption;
            } else if (messageContent.videoMessage?.caption) {
                text = messageContent.videoMessage.caption;
            }

            // Determine media type
            let mediaType = null;
            let mediaUrl = null;
            if (messageContent.imageMessage) {
                mediaType = 'image';
            } else if (messageContent.videoMessage) {
                mediaType = 'video';
            } else if (messageContent.audioMessage) {
                mediaType = 'audio';
            } else if (messageContent.documentMessage) {
                mediaType = 'document';
            }

            // Get sender info
            const isGroup = msg.key.remoteJid.endsWith('@g.us');
            const sender = isGroup
                ? msg.key.participant || msg.key.remoteJid
                : msg.key.remoteJid;

            // Get push name (display name)
            const senderName = msg.pushName || null;

            // Get quoted message ID if this is a reply
            const quotedId = messageContent.extendedTextMessage?.contextInfo?.stanzaId || null;

            const outMsg = {
                type: 'message',
                id: msg.key.id,
                chat: msg.key.remoteJid,
                sender,
                sender_name: senderName,
                timestamp: msg.messageTimestamp,
                text,
                from_me: msg.key.fromMe || false,
                media_type: mediaType,
                media_url: mediaUrl,
                quoted_id: quotedId,
            };

            logger.debug({ msg: outMsg }, 'Received message');
            send(outMsg);
        }
    });

    return sock;
}

// Handle messages from Rust client
async function handleMessage(data) {
    let msg;
    try {
        msg = JSON.parse(data);
    } catch (e) {
        logger.error({ error: e.message }, 'Invalid JSON');
        send({ type: 'error', error: 'Invalid JSON' });
        return;
    }

    logger.debug({ msg }, 'Received command');

    switch (msg.type) {
        case 'init':
            sessionDir = msg.session_dir || '.whatsapp';
            await initWhatsApp();
            break;

        case 'send_message':
            if (!sock) {
                send({ type: 'error', error: 'Not connected', id: msg.id });
                return;
            }
            try {
                const result = await sock.sendMessage(msg.to, { text: msg.text });
                send({ type: 'sent', id: msg.id, message_id: result.key.id });
            } catch (e) {
                logger.error({ error: e.message }, 'Failed to send message');
                send({ type: 'error', error: e.message, id: msg.id });
            }
            break;

        case 'send_media':
            if (!sock) {
                send({ type: 'error', error: 'Not connected', id: msg.id });
                return;
            }
            try {
                let content;
                const mediaBuffer = Buffer.from(msg.data, 'base64');

                switch (msg.media_type) {
                    case 'image':
                        content = { image: mediaBuffer, caption: msg.caption };
                        break;
                    case 'video':
                        content = { video: mediaBuffer, caption: msg.caption };
                        break;
                    case 'audio':
                        content = { audio: mediaBuffer, mimetype: 'audio/mp4' };
                        break;
                    case 'document':
                        content = {
                            document: mediaBuffer,
                            fileName: msg.filename || 'document',
                            caption: msg.caption
                        };
                        break;
                    default:
                        throw new Error(`Unknown media type: ${msg.media_type}`);
                }

                const result = await sock.sendMessage(msg.to, content);
                send({ type: 'sent', id: msg.id, message_id: result.key.id });
            } catch (e) {
                logger.error({ error: e.message }, 'Failed to send media');
                send({ type: 'error', error: e.message, id: msg.id });
            }
            break;

        case 'get_qr':
            // If we have a socket, request will trigger QR
            // Otherwise, need to initialize first
            if (!sock) {
                send({ type: 'error', error: 'Not initialized. Send init first.' });
            }
            break;

        case 'status':
            send({
                type: 'connection',
                status: sock ? 'open' : 'close'
            });
            break;

        case 'disconnect':
            if (sock) {
                await sock.logout();
                sock = null;
            }
            send({ type: 'connection', status: 'close' });
            break;

        default:
            send({ type: 'error', error: `Unknown command: ${msg.type}` });
    }
}

// Start WebSocket server
const wss = new WebSocketServer({ port: PORT });

wss.on('listening', () => {
    logger.info({ port: PORT }, 'WhatsApp bridge WebSocket server started');
    console.log(`WhatsApp bridge listening on ws://localhost:${PORT}`);
});

wss.on('connection', (ws) => {
    logger.info('Rust client connected');

    // Only allow one client
    if (wsClient) {
        wsClient.close();
    }
    wsClient = ws;

    ws.on('message', handleMessage);

    ws.on('close', () => {
        logger.info('Rust client disconnected');
        wsClient = null;
    });

    ws.on('error', (e) => {
        logger.error({ error: e.message }, 'WebSocket error');
    });
});

// Handle process termination
process.on('SIGINT', async () => {
    logger.info('Shutting down...');
    if (sock) {
        await sock.end();
    }
    process.exit(0);
});

process.on('SIGTERM', async () => {
    logger.info('Shutting down...');
    if (sock) {
        await sock.end();
    }
    process.exit(0);
});

File: crates/drbot-whatsapp/src/bridge.rs
//! Bridge protocol for communicating with the Baileys Node.js process.

use serde::{Deserialize, Serialize};

/// Message from Rust to Bridge.
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type")]
pub enum BridgeRequest {
    /// Initialize connection.
    #[serde(rename = "init")]
    Init {
        /// Session data directory.
        session_dir: String,
    },

    /// Send a text message.
    #[serde(rename = "send_message")]
    SendMessage {
        /// Message ID for tracking.
        id: String,
        /// Recipient JID (phone@s.whatsapp.net).
        to: String,
        /// Message text.
        text: String,
    },

    /// Send a media message.
    #[serde(rename = "send_media")]
    SendMedia {
        /// Message ID for tracking.
        id: String,
        /// Recipient JID.
        to: String,
        /// Media type (image, video, audio, document).
        media_type: String,
        /// Base64 encoded media data or URL.
        data: String,
        /// Optional caption.
        caption: Option<String>,
        /// Filename for documents.
        filename: Option<String>,
    },

    /// Get QR code for authentication.
    #[serde(rename = "get_qr")]
    GetQr,

    /// Check connection status.
    #[serde(rename = "status")]
    Status,

    /// Disconnect and cleanup.
    #[serde(rename = "disconnect")]
    Disconnect,
}

/// Message from Bridge to Rust.
#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type")]
pub enum BridgeEvent {
    /// Connection status changed.
    #[serde(rename = "connection")]
    Connection {
        /// Connection state.
        status: ConnectionStatus,
    },

    /// QR code for authentication.
    #[serde(rename = "qr")]
    Qr {
        /// QR code string.
        qr: String,
    },

    /// Message received.
    #[serde(rename = "message")]
    Message {
        /// Message data.
        #[serde(flatten)]
        message: WhatsAppMessage,
    },

    /// Message sent confirmation.
    #[serde(rename = "sent")]
    Sent {
        /// Request ID that was sent.
        id: String,
        /// WhatsApp message ID.
        message_id: String,
    },

    /// Error occurred.
    #[serde(rename = "error")]
    Error {
        /// Error message.
        error: String,
        /// Related request ID if applicable.
        id: Option<String>,
    },

    /// Ready to receive/send messages.
    #[serde(rename = "ready")]
    Ready,
}

/// Connection status.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ConnectionStatus {
    /// Connecting.
    Connecting,
    /// Open and authenticated.
    Open,
    /// Closed.
    Close,
    /// Logged out.
    LoggedOut,
}

/// WhatsApp message from the bridge.
#[derive(Debug, Clone, Deserialize)]
pub struct WhatsAppMessage {
    /// Message ID.
    pub id: String,
    /// Chat JID.
    pub chat: String,
    /// Sender JID.
    pub sender: String,
    /// Sender name (push name).
    pub sender_name: Option<String>,
    /// Message timestamp (Unix seconds).
    pub timestamp: i64,
    /// Text content.
    pub text: Option<String>,
    /// Whether the message is from the user.
    pub from_me: bool,
    /// Media type if present.
    pub media_type: Option<String>,
    /// Media URL if present.
    pub media_url: Option<String>,
    /// Quoted message ID.
    pub quoted_id: Option<String>,
}

impl WhatsAppMessage {
    /// Check if this is a group message.
    pub fn is_group(&self) -> bool {
        self.chat.contains("@g.us")
    }

    /// Get the phone number from a JID.
    pub fn phone_from_jid(jid: &str) -> Option<String> {
        jid.split('@').next().map(|s| s.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bridge_request_serialize() {
        let req = BridgeRequest::SendMessage {
            id: "123".to_string(),
            to: "1234567890@s.whatsapp.net".to_string(),
            text: "Hello".to_string(),
        };
        let json = serde_json::to_string(&req).unwrap();
        assert!(json.contains("send_message"));
        assert!(json.contains("Hello"));
    }

    #[test]
    fn test_bridge_event_deserialize() {
        let json = r#"{"type": "connection", "status": "open"}"#;
        let event: BridgeEvent = serde_json::from_str(json).unwrap();
        match event {
            BridgeEvent::Connection { status } => {
                assert_eq!(status, ConnectionStatus::Open);
            }
            _ => panic!("Wrong event type"),
        }
    }

    #[test]
    fn test_whatsapp_message_deserialize() {
        let json = r#"{
            "id": "MSG123",
            "chat": "1234567890@s.whatsapp.net",
            "sender": "1234567890@s.whatsapp.net",
            "sender_name": "John",
            "timestamp": 1700000000,
            "text": "Hello",
            "from_me": false
        }"#;
        let msg: WhatsAppMessage = serde_json::from_str(json).unwrap();
        assert_eq!(msg.id, "MSG123");
        assert_eq!(msg.text, Some("Hello".to_string()));
        assert!(!msg.is_group());
    }

    #[test]
    fn test_is_group() {
        let msg = WhatsAppMessage {
            id: "1".to_string(),
            chat: "12345678-1234567@g.us".to_string(),
            sender: "1234567890@s.whatsapp.net".to_string(),
            sender_name: None,
            timestamp: 0,
            text: None,
            from_me: false,
            media_type: None,
            media_url: None,
            quoted_id: None,
        };
        assert!(msg.is_group());
    }

    #[test]
    fn test_phone_from_jid() {
        assert_eq!(
            WhatsAppMessage::phone_from_jid("1234567890@s.whatsapp.net"),
            Some("1234567890".to_string())
        );
    }
}
